fetch.bs

```diff
@@ -3834,9 +3834,32 @@ optional <i>CORS-preflight flag</i>, run these steps:
   <p>If <var>request</var>'s <a>service-workers mode</a> is "<code>all</code>", then:
 
   <ol>
+   <li><p>Assert: <var>request</var> 's <a for=request>body</a> is either null or a
+   <a for=/>body</a>.
+
+   <li>
+    <p>Let <var>savedBody</var> be a copy of <var>request</var>'s <a for=request>body</a>.
```

Comment:
What's the benefit of copying the body at this point?

Reply:
Makes sense! Might be worth putting that in a note?

Reply:
I'm not sure I understand. If you clone a request it uses https://fetch.spec.whatwg.org/#concept-body-clone to clone its body, right? This updates body's stream to a non-null value but its source remains the same. Couldn't we always extract from source again?

Reply:
The point I'm trying to make is that it overwrites the request's body's stream and that means request will not be disturbed nor locked. I'd like to fail the request in such a case too (if other conditions are met). This corresponds to "clone-and-ignore" tests in https://github.com/web-platform-tests/wpt/pull/27325.

Reply:
So when you do a `fetch()` with a stream and then the service worker clones the `Request` and doesn't do anything else, the `fetch()` should be rejected? Why?

Reply:
The clone() logic runs in the service worker, so in order to make the request work we need to make the service worker alive until the end-of-stream. That is not desirable, given the developer chose the network fallback option.

Reply:
And the service worker cannot obtain a lock and transfer it?

For the non-service worker case this is happening in https://fetch.spec.whatwg.org/#concept-request-transmit-body which seems to be invoked way too late as it should happen before `fetch()` has returned... This might also be a problem for responses?

Reply:
Sorry, I don't understand the last comment. Can you rephrase it?

Reply:
```js
const req = new Request("about:blank", { method: "POST", body: rs });
fetch(req);
req.text();
```
What causes the last line to throw? As far as I can tell `fetch()` doesn't obtain the reader for _rs_ until way too late.

Reply:
https://fetch.spec.whatwg.org/#fetch-method creates a new Request. That means the 39th step of https://fetch.spec.whatwg.org/#dom-request creates a proxy, which makes `req.body` locked and disturbed.

Reply:
Thanks, I suspect we need to do something similar for passthrough stuff (though probably transfer instead). If the service worker doesn't touch a request with a stream, we do need to transfer it somehow as otherwise the service worker could later try to read from it. I don't think the service worker cloning it but not doing anything else should trigger any weird side effects.

Reply:
Thanks. Does that mean making [handle fetch](https://w3c.github.io/ServiceWorker/#handle-fetch) return a response __or a body__? That is similar to what we are planning to do in our implementation.

Reply:
Yeah, that might be the way to do it. It's a little bit confusing as the body will be of the request so we it'll need some careful notes. It's also only needed when fetch handed the service worker a stream, right? Because only in that case fetch no longer has a body.

Although I suppose if you `clone()` and then read from the non-cloned request we might still want to fail with a network error even if fetch has a body whose source is a string or some such? Bit unclear to me what's best there.

Reply:
My gut feeling is that `req.clone()` should be treated as same as `req.text()`. Hence calling `req.clone()` and returning null in the service worker handler should mean network fallback when the body is null or the body has a non-null source, and that should result in a network error when the body is non-null and has a null source.

Reply:
I disagree. It should be fine to clone a request for monitoring purposes or some such. Cloning should not result in failure all by itself.

Reply:
Hmm, what do you think about my previous comment?

> The clone() logic runs in the service worker, so in order to make the request work we need to make the service worker alive until the end-of-stream. That is not desirable, given the developer chose the network fallback option.

Reply:
I don't understand it. Once fallback is chosen we obtain a lock on the stream (the proxy stuff) as per https://github.com/w3c/ServiceWorker/pull/1563 and the service worker returns null and the stream. Why does the service worker need to remain alive?

Reply:
Let's assume `req` has a body made from a ReadableStream, and it's intercepted in the service worker.

```
self.addEventListener('fetch', (e) => {
  const req = e.request;
  const clonedBody = req.clone();

  ...

  return null;
});
```

The above handler calls `clone()` and returns null. If we support this use case,

 1. [Tee](https://streams.spec.whatwg.org/#rs-tee) runs in the service worker realm,
 2. One of the streams made by [Tee](https://streams.spec.whatwg.org/#rs-tee) is transferred to the network process.

Hence shutting down the service worker shuts down the teeing logic, which breaks the transferred stream in the network process.

Reply:
I'd love some input from @ricea @MattiasBuelens @domenic @jakearchibald. I see the problem and how this might be hard to optimize for, but it also seems very nonintuitive to web developers that merely cloning a request results in a network error. I personally don't think that's acceptable.

Reply:
(For people invited to the discussion, we are talking only about requests with a body made from a ReadableStream. Other requests such as a request with a blob body are out of scope.)

Reply:
I agree that this behaviour is unintuitive, but given that `clone()` uses `Tee()` as its underlying implementation, I can't see another reasonable solution.

Keeping the service worker alive would be bad.

The tee algorithm is running on the service worker thread, so we can't use the new stream. The tee algorithm is tied to the JavaScript realm in which it was created, due to using promises. Even if the stream is transferred away, the algorithm still has to run on the original thread.

The original stream is read from when we call `Tee()` so there's no way to put it back the way it was before the call to `clone()`.

Reply:
It seems that an implementation could use a different implementation though that would avoid keeping the service worker alive. The document that created the request stream (originalStream) is supplying the contents. The service worker gets that stream (requestStream), clones it, keeps one for itself (clonedStream) and hands the other back to the document that created the stream in the first place (requestStream -> originalStreamProxy). It seems an implementation could structure things such that at this point originalStreamProxy gets its contents directly from originalStream and originalStreamProxy gets them from originalStream as well.

I'm not exactly sure where the promises come in here, but it seems that for things that happen in the service worker there is only one realm to chose from and those in the document cannot use the realm from the service worker either.

Reply:
Just to check I'm understanding the above correctly:

1. The _originalRequest_ is sent to the service worker, and `event.request` represents that _originalRequest_.
2. `event.request.clone()` creates two new bodies by teeing, one for `event.request`, and one for the new cloned request.
3. The body associated with _originalRequest_ is now locked, due to the teeing process.

If the above is wrong, it's probably best to ignore the rest of this 😄 

---

We could clone the request (including teeing the body) before sending it to the service worker, giving us _originalRequest_ and _originalRequestForServiceWorker_.

1. _originalRequestForServiceWorker_ is sent to the service worker as `event.request`.
2. If `event.respondWith` is called, then _originalRequest_'s body can be closed. Otherwise, _originalRequestForServiceWorker_'s body can be closed (unless it's already locked).

I think this would allow cloning without increasing memory usage much. It would also mean you can call `event.request.text()` without calling `event.respondWith`, without creating a network error. I don't know if that's a feature or a bug.

Reply:
@jakearchibald I think that would work. 👍

I do agree that we should specify how passing a `Request` object between the original realm and the service worker's realm should work. It seems natural to make `Request` transferrable, or at least make sure that we [transfer the body's stream](https://streams.spec.whatwg.org/#rs-transfer).

Reply:
I think if _originalRequest_ holds a stream, calling `event.request.text()` during the fetch event while not calling `event.repondWith()` should result in a network error still, even if we can make it work due to the additional teeing. It seems nicer to keep enforcing the "read once" rule as much as possible, even though we messed it up for non-stream backed requests in service workers.

Reply:
If we tee the request body before passing it to the service worker (it is hard to implement TBH - I think we will not be able to implement it in the first release), I would like to allow reading the body in the network fallback case. I'm uncomfortable with having magic in the clone operation, and that is more important to me than enforcing the "read once" rule. @annevk, what do you think?

Reply:
Why do yo prefer global "read once" rule to a stream body?

Reply:
To give web developers an API that's as consistent as possible and also give implementations freedom to experiment with various memory optimization techniques.

Reply:
OK, now it clones the original request. I'm going to w3c/ServiceWorker#1563.